collector.c

```diff
@@ -5,301 +5,221 @@
+	 * of PGPROCs iteration under ProcArrayLock and storing results to profile
+	 * and/or history under corresponding another lock.
+	 */
 	LWLockAcquire(ProcArrayLock, LW_SHARED);
-	for (i = 0; i < ProcGlobal->allProcCount; i++)
+	for (int i = 0; i < ProcGlobal->allProcCount; i++)
 	{
-		HistoryItem		item,
-					   *observation;
-		PGPROC		   *proc = &ProcGlobal->allProcs[i];
+		PGPROC	   *proc = GetPGProcByNumber(i);
+		pgwsQueryId queryId = WhetherProfileQueryId ? pgws_proc_queryids[i] : 0;
+		int32 	 	wait_event_info = proc->wait_event_info,
+					pid = proc->pid;
 
-		if (proc->pid == 0)
+		/*
+		 * FIXME: zero pid actually doesn't indicate that process slot is freed.
+		 * After process termination this field becomes unchanged and thereby
+		 * stores the pid of previous process. The possible indicator of process
+		 * termination might be a condition `proc->procLatch->owner_pid == 0`.
+		 * But even in this case ProcArrayLock doesn't protect `owner_pid`
+		 * field from concurrent modifications that might cause race conditions.
+		 *
+		 * Abother option is to use the lists of freed PGPROCs from ProcGlocal:
+		 * freeProcs, walsenderFreeProcs, bgworkerFreeProcs and autovacFreeProcs
+		 * to define indexes of all freed slots in allProcs. But this requires
+		 * acquiring ProcStructLock spinlock that is impractical for iteration
+		 * over so long lists.
+		 *
+		 * The most appropriate solution here is to iterate over ProcArray items
+		 * under ProcArrayLock and over AuxiliaryProcs under ProcStructLock
+		 * spinlock (AuxiliaryProcs contains just NUM_AUXILIARY_PROCS=5 slots)
+		 * or without any locks as it's done in pg_stat_get_activity() function.
+		 * These arrays are not accessible externally and require to add some
+		 * iterator object into corresponding containing modules.
+		 */
+		if (pid == 0)
 			continue;
 
-		if (proc->wait_event_info == 0)
+		// TODO: take into account the state without waiting as CPU time
+		if (wait_event_info == 0)
 			continue;
 
-		/* Collect next wait event sample */
-		item.pid = proc->pid;
-		item.wait_event_info = proc->wait_event_info;
-
-		if (pgws_collector_hdr->profileQueries)
-			item.queryId = pgws_proc_queryids[i];
-		else
-			item.queryId = 0;
-
-		item.ts = ts;
-
 		/* Write to the history if needed */
 		if (write_history)
 		{
-			observation = get_next_observation(observations);
-			*observation = item;
+			int index = pgws_history_ring->index % HistoryBufferSize;
+
+			pgws_history_ring->items[index] = (HistoryItem) {
+				pid, wait_event_info, queryId, GetCurrentTimestamp()
+			};
+			pgws_history_ring->index++;
 		}
 
 		/* Write to the profile if needed */
 		if (write_profile)
 		{
-			ProfileItem	   *profileItem;
-			bool			found;
+			ProfileHashKey		key;
+			ProfileHashEntry   *entry;
 
-			if (!profile_pid)
-				item.pid = 0;
+			/* Set up key for hashtable search */
+			key.pid = WhetherProfilePid ? pid : 0;
```

Comment:
I think we'd better put a memset here:
> Caller must ensure there are no undefined padding bits in the keys!

https://github.com/postgres/postgres/blob/master/src/backend/utils/hash/dynahash.c#L37

We don't have any padding currently, but let's be future proof.

Reply:
Thanks, good addition! Will be added